var/spack/repos/builtin/packages/py-tensorflow/package.py

```diff
@@ -140,38 +146,51 @@ class PyTensorflow(Package, CudaPackage):
+    depends_on('py-absl-py@0.10:0', type=('build', 'run'), when='@2.4.0:2.6')
     depends_on('py-absl-py@0.7.0:', type=('build', 'run'), when='@1.12.1,1.14:2.3')
     depends_on('py-absl-py@0.1.6:', type=('build', 'run'), when='@1.5:1.11')
-    depends_on('py-astunparse@1.6.3:1.6', type=('build', 'run'), when='@2.4.0:')
+    depends_on('py-astunparse@1.6.0:', type=('build', 'run'), when='@2.7:')
+    depends_on('py-astunparse@1.6.3:1.6', type=('build', 'run'), when='@2.4.0:2.6')
     depends_on('py-astunparse@1.6.3', type=('build', 'run'), when='@2.2:2.3')
     depends_on('py-astor@0.6.0:', type=('build', 'run'), when='@1.6:2.1')
     depends_on('py-backports-weakref@1.0:', type=('build', 'run'), when='@1.3: ^python@:3.3')
     depends_on('py-backports-weakref@1.0rc1', type=('build', 'run'), when='@1.2.0:1.2.1')
+    depends_on('py-clang@5.0.0:5.0', type=('build', 'run'), when='@2.6')
+    depends_on('py-libclang@9.0.1:', type=('build', 'run'), when='@2.7:')
     depends_on('py-enum34@1.1.6:', type=('build', 'run'), when='@1.5: ^python@:3.3')
     depends_on('py-enum34@1.1.6:', type=('build', 'run'), when='@1.4.0:1.4.1')
-    depends_on('py-gast@0.3.3', type=('build', 'run'), when='@2.2:')
+    depends_on('py-gast@0.2.1:0.4', type=('build', 'run'), when='@2.7:')
+    depends_on('py-gast@0.4.0', type=('build', 'run'), when='@2.5:2.6')
+    depends_on('py-gast@0.3.3', type=('build', 'run'), when='@2.2:2.4')
     depends_on('py-gast@0.2.2', type=('build', 'run'), when='@1.15:2.1')
     depends_on('py-gast@0.2.0:', type=('build', 'run'), when='@1.6:1.14')
-    depends_on('py-google-pasta@0.2:0', type=('build', 'run'), when='@2.4.0:')
+    depends_on('py-google-pasta@0.1.1:', type=('build', 'run'), when='@2.7:')
+    depends_on('py-google-pasta@0.2:0', type=('build', 'run'), when='@2.4.0:2.6')
     depends_on('py-google-pasta@0.1.8:', type=('build', 'run'), when='@2.1:2.3')
     depends_on('py-google-pasta@0.1.6:', type=('build', 'run'), when='@1.14:2.0')
     depends_on('py-google-pasta@0.1.2:', type=('build', 'run'), when='@1.12.1')
     # propagate the mpi variant setting for h5py/hdf5 to avoid unexpected crashes
-    depends_on('py-h5py@2.10.0:2.10+mpi', type=('build', 'run'), when='@2.2:+mpi')
-    depends_on('py-h5py@2.10.0:2.10~mpi', type=('build', 'run'), when='@2.2:~mpi')
+    depends_on('py-h5py@2.9.0:', type=('build', 'run'), when='@2.7:+mpi')
+    depends_on('py-h5py~mpi@2.9.0:', type=('build', 'run'), when='@2.7:~mpi')
+    depends_on('py-h5py@3.1.0:3.1', type=('build', 'run'), when='@2.5:2.6+mpi')
+    depends_on('py-h5py~mpi@3.1.0:3.1', type=('build', 'run'), when='@2.5:2.6~mpi')
+    depends_on('py-h5py@2.10.0:2.10', type=('build', 'run'), when='@2.2:2.4.999+mpi')
+    depends_on('py-h5py~mpi@2.10.0:2.10', type=('build', 'run'), when='@2.2:2.4.999~mpi')
     depends_on('hdf5+mpi', type='build', when='@2.2:+mpi')
     depends_on('hdf5~mpi', type='build', when='@2.2:~mpi')
```

Comment:
Does TF actually link to hdf5 or does it just use `py-h5py` for everything?

Reply:
TF does not link to HDF5 libraries. I will note that those specs were already there but my guess is they are specified in order to propagate the `mpi` variant. Both the `hdf5` and `py-h5py` packages have `+mpi` set by default, and the `py-h5py` package only sets the `+mpi` variant for `hdf5`, not the `~mpi` variant. So, MPI would be pulled in regardless of the `mpi` variant in TF if the `hdf5` specs were not there. Setting the `hdf5` variants explicitly allows `mpi` to be toggled and keep the `mpi` variant in sync across the packages. Figuring out what variants work together is something that could be further explored, but I do not think this PR is the correct vehicle for that.